### PR TITLE
🐛 Fix `find_from_id_or_value` returning wrong token

### PR DESCRIPTION
**Pull request by josemigallas**
[THREESCALE-16355: AccessToken.find_from_id_or_value incorrectly matches hex token values as integer IDs due to MySQL silent casting](https://redhat.atlassian.net/browse/THREESCALE-16355)

MySQL silently casts strings to integers when comparing against integer columns ([docs](https://dev.mysql.com/doc/refman/8.0/en/type-conversion.html)). When a token value like `"1234567eab..."` is passed to `find_by(id:)`, MySQL casts it to `1234567`, potentially matching a different token by id.

This bug was identified after investigating flaky test `AccessTokenTest#test_find_from_id_or_value_and_bang`, which from time to time runs into this edge case (presumably 1/16 of times). In a real production environment it is very unlikely that we will be hit for 2 reasons:
  1. Long ids makes it unlikely to match the leading digits of another token's hex value
  2. Tokens are scoped to `current_user`, it's only possible for a user to accidentally get/update/delete their own tokens.
However, the bug is real and it 

The solution I propose is to **route by length**. Token values are always 64 chars (legacy) or 96 chars (new SHA384), while bigint ids are at most 19 digits. Inputs of 19 chars or fewer go to `find_by(id:)`, longer inputs go to `find_from_value`. Zero extra cost in either path. 

**Alternatives considered**:
- **Switch lookup order** (`find_from_value` first): simpler one-line change but adds an extra SHA384 hash + DB query on every id lookup, for little benefit.
- **Fix only the flaky test**: make it deterministic without changing the code. Hides the bug under the rug, even though it's very rare that it would cause issues in production.

### CODE REVIEW

File: app/models/access_token.rb [hunk 1]
```diff
@@ -129,7 +129,11 @@ def self.find_from_id_or_value!(id_or_value)
   end
 
   def self.find_from_id_or_value(id_or_value)
-    find_by(id: id_or_value) || find_from_value(id_or_value)
+    if id_or_value.to_s.length <= 19
+      find_by(id: id_or_value)
+    else
+      find_from_value(id_or_value)
+    end
```
Comment (jlledom) [inline app/models/access_token.rb:136]: I think this is fragile because there might be old access tokens in DB which plan text value is less than 19 characters long, e.g. `"secret"`. We can't know which kind of data is in DB so better assume the worst.

Instead, I think it's better to check whether the given string is in fact all digits, in that case make it go through the `id` route

```suggestion
    if id_or_value.to_s.match?(/\A\d+\z/)
      find_by(id: id_or_value)
    else
      find_from_value(id_or_value)
    end
```

The only counter-example for this is when a generated access token is legitimately composed by a 96-char sequence of only numbers, which is very unlikely. In particular, according to AI, is this unlikely:

```
It has probability ~10^-20. To put that in perspective: if you generated one
new access token every second, you'd expect to wait about 1.3 trillion years
before seeing an all-digit value — roughly 95× the age of the universe.
```

An even in that case the only consequence would be wrong 404.

Comment (akostadinov) [reply app/models/access_token.rb:136]: We can also add the length limit to the RE. But why not keep the fallback as well? Doesn't seem to complicate things much more.


```suggestion
    # MySQL casts `123foo` to `123` which sometimes causes test failures
    if id_or_value.to_s.match?(/\A\d{,19}\z/)
      found = find_by(id: id_or_value)
      return found if found
    end
    
    find_from_value(id_or_value)
```

Comment (jlledom) [reply app/models/access_token.rb]: yeah @akostadinov that's better.

Comment (josemigallas) [reply app/models/access_token.rb]: @akostadinov that regex was in fact what Claude suggested me first but I thought it would change things too much.

If y'all agree I will use this.

File: test/models/access_token_test.rb [hunk 2]
```diff
@@ -95,6 +95,22 @@ def test_find_from_id_or_value_and_bang
     assert_raise(ActiveRecord::RecordNotFound) { AccessToken.find_from_id_or_value!('fake') }
   end
 
+  # Regression test for https://redhat.atlassian.net/browse/THREESCALE-16355
+  # When the leading digits of a token's value match another token's id, MySQL silently casts the
+  # string to an integer. find_from_value must run first to avoid returning the wrong token.
+  def test_find_from_id_or_value_edge_cases
```
Comment (jlledom) [inline test/models/access_token_test.rb:102]: If you accept the suggestions in the `find_from_id_or_value` method, then this tests becomes obsolete.

Also, there's no need to check fro 64-byte tokens because we migrated the table and now all tokens are 96-byte.

File: app/models/access_token.rb (hunk 1 above)
Comment (akostadinov) [reply app/models/access_token.rb]: > that regex was in fact what Claude suggested me first

am I starting to think like AI :grimacing:

File: test/models/access_token_test.rb (hunk 2 above)
Comment (josemigallas) [reply test/models/access_token_test.rb:102]: I don't see why it's obsolete, it's a regression test for THREESCALE-16355. The tokens length are incidental.

**Review by akostadinov — APPROVED** [review]

**Review by jlledom — APPROVED** [review]
